thirdparty/github.go

```diff
@@ -891,58 +767,31 @@ func GithubAuthenticate(ctx context.Context, code, clientId, clientSecret string
 
 // GetTaggedCommitFromGithub gets the commit SHA for the given tag name.
 func GetTaggedCommitFromGithub(ctx context.Context, token, owner, repo, tag string) (string, error) {
-	installationToken, err := GetInstallationToken(ctx, owner, repo, nil)
-	if err != nil {
-		ctx, httpClient, putClient := getGithubClient(ctx, installationToken, "GetTaggedCommitFromGithub", retryConfig{retry: true}, []attribute.KeyValue{
-			attribute.String(githubOwnerAttribute, owner),
-			attribute.String(githubRepoAttribute, repo),
-			attribute.String(githubRefAttribute, tag),
-		})
-		defer putClient()
-		client := github.NewClient(httpClient)
-
-		tags, resp, err := client.Repositories.ListTags(ctx, owner, repo, nil)
-		if resp != nil {
-			defer resp.Body.Close()
-		}
-
-		for _, t := range tags {
-			if t.GetName() == tag {
-				return t.GetCommit().GetSHA(), nil
-			}
-		}
-
-		grip.Debug(message.WrapError(err, message.Fields{
-			"ticket":  "EVG-19966",
-			"message": "failed to get tagged commit from github",
-			"caller":  "GetTaggedCommitFromGithub",
-			"owner":   owner,
-			"repo":    repo,
-			"resp":    resp,
-		}))
-	}
-
-	// Fallback to not using the GitHub app on error.
-	ctx, legacyClient, putLegacyClient := getGithubClient(ctx, token, "GetTaggedCommitFromGithub", retryConfig{retry: true}, []attribute.KeyValue{
+	caller := "GetTaggedCommitFromGithub"
+	ctx, span := tracer.Start(ctx, caller, trace.WithAttributes(
+		attribute.String(githubEndpointAttribute, caller),
 		attribute.String(githubOwnerAttribute, owner),
 		attribute.String(githubRepoAttribute, repo),
 		attribute.String(githubRefAttribute, tag),
-	})
-	defer putLegacyClient()
+	))
+	defer span.End()
+
+	githubClient := getGithubClient(getInstallationToken(ctx, token, owner, repo, nil), caller, retryConfig{retry: true})
 
 	url := fmt.Sprintf("https://api.github.com/repos/%s/%s/git/refs/tags/%s", owner, repo, tag)
-	newResp, err := legacyClient.Get(url)
-	if newResp != nil {
-		defer newResp.Body.Close()
+	resp, err := githubClient.Client().Get(url)
+	if resp != nil {
+		defer resp.Body.Close()
+		span.SetAttributes(attribute.Bool(githubCachedAttribute, respFromCache(resp)))
 	}
 	if err != nil {
 		return "", errors.Wrap(err, "failed to get tag information from GitHub")
 	}
-	if newResp == nil {
+	if resp == nil {
 		return "", errors.New("invalid github response")
 	}
 
-	respBody, err := io.ReadAll(newResp.Body)
+	respBody, err := io.ReadAll(resp.Body)
```

Comment:
I see the old code looping through a list of tags while this one seems to parse only one tag. However, the url seems to be the same. Where is the discrepancy from?

Reply:
I'm not sure why we were doing it differently with the app than with the legacy token. To me it seems fraught to list all the tags on the repository since the beginning of time. Also, I think if we wanted to do that we'd need to use the options to get all the pages ([the docs](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags) say we'll get a max of 100 at a time).

@bynn was it intentional to have them use different endpoints?

Reply:
the "https://api.github.com/repos/%s/%s/git/refs/tags/%s" doesn't seem to exist on the client that we're using. 

It is technically the same route but without the "tags/{tag}" part at the end. It probably makes more sense to call it directly using the route since using the client will make us do extra work

Reply:
I'd originally thought about switching to [this route](https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#get-a-reference) to get the SHA for the tag (because it's documented and [implemented by the SDK](https://github.com/google/go-github/blob/ae774d0751d40497e72d1ef2e3b98160adffd93e/github/git_refs.go#L50-L68)) but decided to not change unrelated things.